File: conformance/code/accesscontrol/internal/model/nullable.go
package model

import "encoding/json"

// Nullable wraps an optional PATCH field so we can distinguish all three
// caller intents that JSON otherwise conflates:
//
//   - Field absent from the request body   → Set=false   (don't touch the column)
//   - Field present as JSON null           → Set=true,  Null=true   (clear the column)
//   - Field present with a value           → Set=true,  Null=false, Value=<v>
//
// Standard Go pointer fields cannot express the second state — both "absent"
// and "null" decode to nil. Nullable[T] uses a custom UnmarshalJSON to record
// presence at decode time, before any null collapsing happens.
type Nullable[T any] struct {
	Set   bool
	Null  bool
	Value T
}

// UnmarshalJSON is invoked only when the field is present in the JSON body.
// Therefore Set is always true here. We then check whether the literal `null`
// was sent and either record that or decode the inner value.
func (n *Nullable[T]) UnmarshalJSON(data []byte) error {
	n.Set = true
	if string(data) == "null" {
		n.Null = true
		return nil
	}
	return json.Unmarshal(data, &n.Value)
}

File: conformance/code/accesscontrol/src/main/java/com/digit/accesscontrol/model/Nullable.java
package com.digit.accesscontrol.model;

import tools.jackson.databind.JsonNode;

/**
 * Wraps an optional PATCH field so we can distinguish the three caller intents JSON conflates:
 * <ul>
 *   <li>Field absent          → set=false (don't touch the column)</li>
 *   <li>Field present as null  → set=true, isNull=true (clear the column)</li>
 *   <li>Field present w/ value → set=true, isNull=false, value=&lt;v&gt;</li>
 * </ul>
 * Mirrors Go internal/model/nullable.go (Nullable[T]). Presence/null are populated by the
 * controller from the raw JSON tree before the typed read, since Jackson cannot itself distinguish
 * "absent" from "null" for a default-constructed field.
 */
public class Nullable<T> {
    private boolean set;
    private boolean isNull;
    private T value;

    public boolean isSet() { return set; }
    public void setSet(boolean set) { this.set = set; }
    public boolean isNull() { return isNull; }
    public void setNull(boolean aNull) { this.isNull = aNull; }
    public T getValue() { return value; }
    public void setValue(T value) { this.value = value; }

    /**
     * Populates this wrapper from a raw JSON tree node for the given field, mirroring the Go
     * UnmarshalJSON semantics. {@code node} is the field's value node, or null when the field is
     * absent from the body.
     */
    public void populate(JsonNode node, java.util.function.Function<JsonNode, T> valueExtractor) {
        if (node == null) {
            this.set = false;
            return;
        }
        this.set = true;
        if (node.isNull()) {
            this.isNull = true;
            return;
        }
        this.value = valueExtractor.apply(node);
    }
}
